File: app/lib/core/utils/order_totals.dart
import 'package:decimal/decimal.dart';

import '../errors/failures.dart';

Decimal money(String value) {
  final parsed = parseMoneyInput(value);
  if (parsed == null) {
    throw FormatException('Invalid amount: $value');
  }
  return parsed;
}

/// Parses currency/quantity text (`10`, `+10`, `-2.5`). Returns null if invalid.
Decimal? parseMoneyInput(String raw) {
  var text = raw.trim().replaceAll(',', '');
  if (text.isEmpty) return null;
  if (text.startsWith('+')) {
    text = text.substring(1).trim();
    if (text.isEmpty) return null;
  }
  try {
    return Decimal.parse(text);
  } catch (_) {
    return null;
  }
}

Decimal parseMoneyInputOrThrow(String raw, {String label = 'Amount'}) {
  final parsed = parseMoneyInput(raw);
  if (parsed == null) {
    throw Failure('$label is not a valid number.');
  }
  return parsed;
}

String moneyString(Decimal value) => value.round(scale: 2).toStringAsFixed(2);

class OrderLineInput {
  const OrderLineInput({
    required this.quantity,
    required this.unitPrice,
    this.extraPrice = '0',
    this.discountAmount = '0',
  });

  final String quantity;
  final String unitPrice;
  final String extraPrice;
  final String discountAmount;
}

class OrderTotals {
  const OrderTotals({
    required this.subtotal,
    required this.discountAmount,
    required this.taxAmount,
    required this.total,
    required this.lineTotals,
  });

  final String subtotal;
  final String discountAmount;
  final String taxAmount;
  final String total;
  final List<String> lineTotals;
}

OrderTotals calculateOrderTotals({
  required List<OrderLineInput> lines,
  String orderDiscountAmount = '0',
  String taxRate = '0',
}) {
  final lineTotals = <Decimal>[];
  for (final line in lines) {
    final qty = money(line.quantity);
    final unit = money(line.unitPrice) + money(line.extraPrice);
    final raw = unit * qty;
    final discounted = raw - money(line.discountAmount);
    lineTotals.add(discounted < Decimal.zero ? Decimal.zero : discounted);
  }
  final subtotal = lineTotals.fold(Decimal.zero, (sum, value) => sum + value);
  var discount = money(orderDiscountAmount);
  if (discount > subtotal) {
    discount = subtotal;
  }
  final taxable = subtotal - discount;
  final taxAmount = (taxable * money(taxRate)).round(scale: 2);
  final total = taxable + taxAmount;
  return OrderTotals(
    subtotal: moneyString(subtotal),
    discountAmount: moneyString(discount),
    taxAmount: moneyString(taxAmount),
    total: moneyString(total),
    lineTotals: lineTotals.map(moneyString).toList(),
  );
}

File: backend/src/modules/orders/orderTotals.ts
import { Decimal } from 'decimal.js';
import { money, moneyString } from '../../utils/money';

export interface OrderLineInput {
  quantity: string | number;
  unitPrice: string | number;
  extraPrice?: string | number;
  discountAmount?: string | number;
}

export interface OrderTotalsInput {
  lines: OrderLineInput[];
  orderDiscountAmount?: string | number;
  taxRate: string | number;
}

export interface OrderTotals {
  subtotal: string;
  discountAmount: string;
  taxAmount: string;
  total: string;
  lineTotals: string[];
}

export function calculateOrderTotals(input: OrderTotalsInput): OrderTotals {
  const lineTotals = input.lines.map((line) => {
    const qty = money(line.quantity);
    if (qty.lte(0)) {
      throw new Error('INVALID_QUANTITY');
    }
    const unit = money(line.unitPrice).plus(money(line.extraPrice ?? 0));
    const raw = unit.times(qty);
    const discount = money(line.discountAmount ?? 0);
    const lineTotal = Decimal.max(raw.minus(discount), money(0));
    return lineTotal;
  });

  const subtotal = lineTotals.reduce((sum, value) => sum.plus(value), money(0));
  const discountAmount = Decimal.min(money(input.orderDiscountAmount ?? 0), subtotal);
  const taxable = subtotal.minus(discountAmount);
  const taxAmount = taxable.times(money(input.taxRate));
  const total = taxable.plus(taxAmount);

  return {
    subtotal: moneyString(subtotal),
    discountAmount: moneyString(discountAmount),
    taxAmount: moneyString(taxAmount),
    total: moneyString(total),
    lineTotals: lineTotals.map((value) => moneyString(value)),
  };
}
